h/models/group.py

```diff
@@ -49,6 +65,21 @@ class Group(Base, mixins.Timestamps):
 
     description = sa.Column(sa.UnicodeText())
 
+    #: Which type of user is allowed to join this group, possible values are:
+    #: authority, None
+    joinable_by = sa.Column(sa.Enum(JoinableBy, name='group_joinable_by'),
+                            nullable=True)
+
+    #: Which type of user is allowed to read annotations in this group,
+    #: possible values are: authority, members, world
+    readable_by = sa.Column(sa.Enum(ReadableBy, name='group_readable_by'),
+                            nullable=True)
+
+    #: Which type of user is allowed to write to this group, possible values
+    #: are: authority, members
+    writeable_by = sa.Column(sa.Enum(WriteableBy, name='group_writeable_by'),
+                             nullable=True)
```

Comment:
Do you think it might be better for data integrity for `joinable_by`, `readable_by` and `writeable_by` to be Python `@property`s that aren't stored in the db but are just implemented in Python and return different values depending on a `type` field that is stored in the db? And `type` would be an enum that, for now, has I guess only one possible value `"private"` or maybe two `"private"` and `"publisher"`? And of course if we want to make use of Pyramid's ACL system `Group` can have an `__acl__()` method that works based on these and any other properties of the group (like `memex.models.Annotation.__acl__()`).

I'm thinking some of the advantages might be that, like with the current `Annotation` model, it won't be possible at the database level for an invalid value to be stored (for example a group that is joinable by no one but only readable by its members etc). We can guarantee that, based on the more restricted properties that are actually stored in the db, the `__acl__()` method will only generate the valid possible ACLs and not any contradictory ones. And it'll be easier to look in the db or model and find all the "private" groups or all the "publisher" groups by just looking at the `type` column. Adding a new type of group would just be a case of adding a new value to the type column's enum and extending the `__acl__()` method to support it.